practical/ProcessMining/group1/task2/alphaminer.py

```diff
@@ -0,0 +1,521 @@
+
+    def footprint_matrix(self) -> pd.DataFrame:
+        """
+        Generates the footprint matrix for the process.
+
+        Returns:
+            pd.DataFrame: The footprint matrix.
+        """
+        matrix = pd.DataFrame(index=sorted(self.activities.values()), columns=sorted(self.activities.values()))
+
+        for pair in self.all_pairs:
+            a1, a2 = pair
+            a1_value, a2_value = self.activities[a1], self.activities[a2]
+            if any(np.array_equal(pair, p) for p in self.parallel_pairs):
+                matrix.at[a1_value, a2_value] = '||'
+            elif any(np.array_equal(pair, p) for p in self.sequential_pairs):
+                matrix.at[a1_value, a2_value] = '→'
+            elif any(np.array_equal(pair, p) for p in self.not_following_pairs):
+                matrix.at[a1_value, a2_value] = '#'
+            elif any(np.array_equal(pair, p) for p in self.before_pairs):
+                matrix.at[a1_value, a2_value] = '←'
+            else:
+                matrix.at[a1_value, a2_value] = ''
+
+        return matrix
+
+    def _import_event_log(self, file_path: str, case_id='case_id', activity_key='activity',
+                          timestamp_key='timestamp') -> Tuple[pd.DataFrame, Dict[int, str], List[Tuple[int, int]]]:
+        """
+        Imports the event log from a file.
+
+        Parameters:
+            file_path (str): The path to the event log file.
+
+        Returns:
+            Tuple[pd.DataFrame, Dict[int, str], List[Tuple[int, int]]]: The event log data,
+            the mapping of activity IDs to activity names, and all pairs of activities.
+        """
+        if not os.path.exists(file_path):
+            raise Exception("File does not exist")
+
+        extension = os.path.splitext(file_path)[1]
+        if extension == '.csv':
+            event_log = pd.read_csv(file_path, sep=';')
+            event_log = pm4py.format_dataframe(event_log, case_id=case_id, activity_key=activity_key,
+                                               timestamp_key=timestamp_key)
+        elif extension == '.xes':
+            event_log = pm4py.read_xes(file_path)
+        else:
+            raise Exception("File extension must be .csv or xes")
+
+        event_log = event_log.sort_values(['case:concept:name', 'time:timestamp'])
+        event_log = (event_log[["case:concept:name", "concept:name"]]
+                     .rename(columns={"case:concept:name": "case_id", "concept:name": "activity"}))
+        activities = self._create_alphabet(event_log)
+        all_pairs = [(a1, a2) for a1 in activities.keys() for a2 in activities.keys()]
+        event_log['activity_id'] = event_log.apply(lambda row: dict(zip(activities.values(),
+                                                                        activities.keys())).get(row['activity']), 1)
+        return event_log, activities, all_pairs
+
+    def _create_alphabet(self, event_log: pd.DataFrame) -> Dict[int, str]:
```

Comment:
Tiny remark, is "alphabet" good naming for all cases? If there are more activities than letters in the latin alphabet then this naming might be slighly off

Reply:
I was thinking about [alphabet in the line of formal languages](https://en.wikipedia.org/wiki/Alphabet_(formal_languages)) when naming that, but I see that this could be confusing for the dictionary of unique activities. Maybe `unique_activities` or something like that would fit better.